src/poly/evaluator.rs

```diff
@@ -0,0 +1,542 @@
+    ) -> Polynomial<F, Self>;
+}
+
+impl BasisOps for Coeff {
+    fn empty_poly<F: FieldExt>(domain: &EvaluationDomain<F>) -> Polynomial<F, Self> {
+        domain.empty_coeff()
+    }
+
+    fn constant_term<F: FieldExt>(chunk_size: usize, chunk_index: usize, scalar: F) -> Vec<F> {
+        let mut chunk = vec![F::zero(); chunk_size];
+        if chunk_index == 0 {
+            chunk[0] = scalar;
+        }
+        chunk
+    }
+
+    fn linear_term<F: FieldExt>(
+        _: &EvaluationDomain<F>,
+        chunk_size: usize,
+        chunk_index: usize,
+        scalar: F,
+    ) -> Vec<F> {
+        let mut chunk = vec![F::zero(); chunk_size];
+        // If the chunk size is 1 (e.g. if we have a small k and many threads), then the
+        // linear coefficient is the second chunk. Otherwise, the chunk size is greater
+        // than one, and the linear coefficient is the second element of the first chunk.
+        if chunk_size == 1 && chunk_index == 1 {
+            chunk[0] = scalar;
+        } else if chunk_index == 0 {
+            chunk[1] = scalar;
+        }
+        chunk
+    }
+
+    fn rotate<F: FieldExt>(
+        _: &EvaluationDomain<F>,
+        _: &Polynomial<F, Self>,
+        _: Rotation,
+    ) -> Polynomial<F, Self> {
+        panic!("Can't rotate polynomials in the standard basis")
+    }
+}
+
+impl BasisOps for LagrangeCoeff {
+    fn empty_poly<F: FieldExt>(domain: &EvaluationDomain<F>) -> Polynomial<F, Self> {
+        domain.empty_lagrange()
+    }
+
+    fn constant_term<F: FieldExt>(chunk_size: usize, _: usize, scalar: F) -> Vec<F> {
+        vec![scalar; chunk_size]
+    }
+
+    fn linear_term<F: FieldExt>(
+        domain: &EvaluationDomain<F>,
+        chunk_size: usize,
+        chunk_index: usize,
+        scalar: F,
+    ) -> Vec<F> {
+        // Take every power of omega within the chunk, and multiply by scalar.
+        let omega = domain.get_omega();
+        let start = chunk_size * chunk_index;
+        (0..chunk_size)
+            .scan(omega.pow_vartime(&[start as u64]), |cur_omega, _| {
+                let ret = *cur_omega * scalar;
+                *cur_omega *= omega;
+                Some(ret)
+            })
+            .collect()
+    }
+
+    fn rotate<F: FieldExt>(
+        _: &EvaluationDomain<F>,
+        poly: &Polynomial<F, Self>,
+        rotation: Rotation,
+    ) -> Polynomial<F, Self> {
+        poly.rotate(rotation)
+    }
+}
+
+impl BasisOps for ExtendedLagrangeCoeff {
+    fn empty_poly<F: FieldExt>(domain: &EvaluationDomain<F>) -> Polynomial<F, Self> {
+        domain.empty_extended()
+    }
+
+    fn constant_term<F: FieldExt>(chunk_size: usize, _: usize, scalar: F) -> Vec<F> {
+        vec![scalar; chunk_size]
+    }
+
+    fn linear_term<F: FieldExt>(
+        domain: &EvaluationDomain<F>,
+        chunk_size: usize,
+        chunk_index: usize,
+        scalar: F,
+    ) -> Vec<F> {
+        // Take every power of the extended omega within the chunk, and multiply by scalar.
```

Comment:
Is this supposed to be omega^i * zeta * scalar, or (omega * zeta)^i * scalar?

Reply:
Here is what's going on here: we need the polynomial `p(X) = scalar X` evaluated over the coset of the extended domain, which means we need `p(\zeta X)` evaluated over the extended domain (`\zeta` generates the coset), which means we need each `scalar * \zeta * extended_omega^i`.